src/renderer/opengl/linux/linux_gl_context.cpp

```diff
@@ -0,0 +1,196 @@
+                    bestFbc = i;
+                    bestNumSamp = samples;
+                }
+            }
+            XFree(visual);
+        }
+        GLXFBConfig bestFbConfig = fbConfig[bestFbc];
+        XFree(fbConfig);
+        fbConfig = nullptr;
+
+        return bestFbConfig;
+    }
+
+    static constexpr auto glxMajorVersion = 1u;
+    static constexpr auto glxMinorVersion = 4u;
+
+    ::Window window{0};
+    XVisualInfo *visual{nullptr};
+    Display *display{nullptr};
+    Screen *screen{nullptr};
+    GLXFBConfig frameBufferConfig{nullptr};
+    GLXContext handleRenderingContext{nullptr};
+    XID screenId;
+};
+
+GLContext::GLContext() : m_impl{std::make_unique<Impl>()} {
+    m_impl->display = XOpenDisplay(static_cast<const char *>(nullptr));
+    NOX_ASSERT_MSG(m_impl->display == nullptr, "Unable to connect to X11 display");
+
+    m_impl->screen = DefaultScreenOfDisplay(m_impl->display); // NOLINT
+    m_impl->screenId = DefaultScreen(m_impl->display);
+
+    {
+        gladLoaderLoadGLX(m_impl->display, m_impl->screenId);
+        int32_t majorGLX{}, minorGLX{};
+        glXQueryVersion(m_impl->display, &majorGLX, &minorGLX);
+        NOX_ASSERT_MSG((majorGLX <= m_impl->glxMajorVersion) && (minorGLX < m_impl->glxMinorVersion), "This system doesn't support GLX 1.4");
+
+        NOX_LOG_TRACE(OPENGL, "GLX client version: {}", glXGetClientString(m_impl->display, GLX_VERSION));
+        NOX_LOG_TRACE(OPENGL, "GLX client vendor: {}", glXGetClientString(m_impl->display, GLX_VENDOR));
+        NOX_LOG_TRACE(OPENGL, "GLX server version: {}", glXQueryServerString(m_impl->display, m_impl->screenId, GLX_VERSION));
+        NOX_LOG_TRACE(OPENGL, "GLX server vendoe: {}", glXQueryServerString(m_impl->display, m_impl->screenId, GLX_VENDOR));
+    }
+
+    // creating dummy opengl context for early gl initialization
+    const PixelFormatDescriptor defaultPixelFormatDescriptor{};
+    m_impl->setContextPixelFormatAndVisual(defaultPixelFormatDescriptor);
+
+    GLXContext dummyContext = glXCreateContext(m_impl->display, m_impl->visual, NULL, GL_TRUE);
+    auto result = glXMakeCurrent(m_impl->display, 0, dummyContext);
+    NOX_ASSERT_MSG(!result, "Unable to make OpenGL rendering dummy context current");
+
+    {
+        gladLoaderLoadGL();
+        GLint majorVersion, minorVersion;
+        glGetIntegerv(GL_MAJOR_VERSION, &majorVersion);
+        glGetIntegerv(GL_MINOR_VERSION, &minorVersion);
+        NOX_ASSERT_MSG((majorVersion != glMajorVersion) || (minorVersion != glMinorVersion), "This system doesn't support OpenGL 4.6");
+    }
+    glXDestroyContext(m_impl->display, dummyContext);
+
+    XCloseDisplay(m_impl->display);
+    m_impl->display = nullptr;
+    m_impl->frameBufferConfig = nullptr;
+    m_impl->screen = nullptr;
+    m_impl->visual = nullptr;
+}
+
+GLContext::~GLContext() {
+    glXDestroyContext(m_impl->display, m_impl->handleRenderingContext);
```

Comment:
If `SwapChain` isn't created before `Renderer` is destroyed this will fail as `m_impl->display` and `m_impl->handleRenderingContext` will be nullptr.